crypto.go

```diff
@@ -47,22 +57,61 @@ func (s dhkemScheme) Marshal(pk KEMPublicKey) []byte {
 	return s.group.Marshal(pk)
 }
 
+func (s dhkemScheme) MarshalPrivate(sk KEMPrivateKey) []byte {
+	return s.group.MarshalPrivate(sk)
+}
+
 func (s dhkemScheme) Unmarshal(enc []byte) (KEMPublicKey, error) {
 	return s.group.Unmarshal(enc)
 }
 
-func (s dhkemScheme) Encap(rand io.Reader, pkR KEMPublicKey) ([]byte, []byte, error) {
-	skE, pkE, err := s.group.GenerateKeyPair(rand)
+func (s dhkemScheme) UnmarshalPrivate(enc []byte) (KEMPrivateKey, error) {
+	return s.group.UnmarshalPrivate(enc)
+}
+
+func (s dhkemScheme) setEphemeralKeyPair(sk KEMPrivateKey, pk KEMPublicKey) {
+	s.skE = sk
+	s.pkE = pk
+}
+
+func (s dhkemScheme) ephemeralPublic() KEMPublicKey {
+	return s.pkE
+}
+
+func (s dhkemScheme) ephemeralPrivate() KEMPrivateKey {
+	return s.skE
+}
+
+func (s dhkemScheme) getEphemeralKeyPair(rand io.Reader) (KEMPrivateKey, KEMPublicKey, error) {
+	var skE KEMPrivateKey
+	var pkE KEMPublicKey
+	var err error
+
+	if s.skE != nil && s.pkE != nil {
+		skE = s.skE
+		pkE = s.pkE
+	} else {
+		skE, pkE, err = s.group.GenerateKeyPair(rand)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return skE, pkE, nil
+}
+
+func (s dhkemScheme) Encap(rand io.Reader, pkR KEMPublicKey) ([]byte, []byte, KEMPrivateKey, KEMPublicKey, error) {
```

Comment:
You don't need to return the public key here, since you're already returning it in marshaled form.